pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -308,3 +308,35 @@ def is_module(self) -> bool:
         from pytorch_lightning import LightningDataModule, LightningModule  # prevent cyclic import
 
         return isinstance(self.instance, (LightningModule, LightningDataModule))
+
+
+@dataclass
+class _DataHookSource:
```

Comment:
I disagree with this. The name and usage is inspired by the class I wrote above, but this not how it is meant to work. 
A XSource is meant to declare one source and one way to retrieve an instance from that source. 
As you can see in the design above, the source only contains one reference and what that reference is, is determined before instantiating the source.

It is unclear to me why we need two classes instead of just one.

Reply:
Given the name _DataHookSource, perhaps your intention was to scrap the _DataLoaderSource above? After all, the methods x_dataloader() are hooks too.

Reply:
yes but those hooks are more like data providers and the data source can also be explicit too (`.fit(train_dataloader=...`), but these hooks are sort of like data-handlers. For BC and the possibility of overriding the same method in both module and datamodule, I thought of creating an additional class that can work with any hooks that are common in both. For eg, if users provide datamodule but override one of these hooks within LightningModule, then we can't rely on just the datamodule and ignore the implementation within LightningModule.

One possible improvement I can think of is we can start raising a warning there if the data provider is different from the component where the hook is overridden and later merge both of them. Also possibly create a temp datamodule for explicit dataloaders as suggested here: https://github.com/PyTorchLightning/pytorch-lightning/issues/10498#issuecomment-973812617

Reply:
Perhaps `DataHookSelector` would be a better name?